active/0030-emqx_linear_channel_registry.md

```diff
@@ -0,0 +1,321 @@
+   
+   The living node may remove the channel belonging to the down node. It must be done only once within 
+   the cluster successfully.
+   
+   For replicant node down, its upstream core node should clean the channels for it.
+   For core node down, @TODO
+   
+   
+### Drawbacks
+
+- Compare to the `ekka_locker`, using transaction will stress the `mnesia_locker` which is single point of bottleneck.
+
+  But `mnesia_locker` is more robust than `ekka_locker`.
+
+- "registration history function" will not be supported as we don't want large data set.
+  It must be reimplemented.
+   
+## Backwards Compatibility
+
+`LCR` should be disabled by default with a feature switch.
+
+### LCR Disabled
+
+When `LCR` is disabled, there should be no backwards compatibility issue, and the EMQX cluster should work as it does in the old version.
+
+### LCR Enabled within the cluster
+
+Once `LCR` is enabled, `LCR` can co-exist with `emqx_cm_registry` with a feature switch.
+
+READs from `LCR` storage only. WRITES go to both if `emqx_cm_registry` is also enabled (global), but DO NOT use `ekka_locker:trans`.
+
+
+### LCR partially Enabled within the cluster
+
+During a rolling upgrade, the EMQX cluster could run with `LCR` partially enabled due to the mixed versions of EMQX nodes.
+
+To be compatible with old bpapi calls from the old version of EMQX, there are two problems to solve:
+
+1.  In the RPC call, no channel version is missing.
+2.  `emqx_cm_registry` is the only provider, and it has no channel vsn.
+
+Thus, we define the fallback method here for Node evacuation and rolling upgrade:
+
+For newer EMQX handling the call from older EMQX, the channel version uses the value of the timestamp: `disconnected_at`.
+
+Newer EMQX MUST NOT disable the `emqx_cm_registry`; as part of the standard implementation, the writes will go to both storage.
+
+For newer EMQX calling the older EMQX, the newer EMQX should write to the `LCR` once the bpapi call is successful. The newer EMQX MUST not update the `emqx_cm_registry` storage as it is part of the call handling on the older EMQX.
+
+
+### LCR runtime enabling
+
+UNSUPPORTED for now.
+
+### LCR runtime disabling
+
+Supported, but we don't know the side effects. Leave it as @TODO.
+
+
+## Document Changes
+
+If there is any document change, give a brief description of it here.
+
+## Testing Suggestions
+
+### Functional test
+
+1.  The existing common test suites should be expanded with a new group `lcr_on`, `lcr_off`
+
+    -   takeover suite
+    -   cm suite
+
+    Add a new cluster test suite for testing core + replicant roles covering known issues in previous chapters
+    
+2.  Rolling upgrade should be tested with the `LCR` switch on.
+
+## Performance test
+
+Performance tests should be performed against a 3 cores + 3 replicants cluster.
+
+Compare the performance with `LCR` on/off with the following scenarios:
+
+1.  initial connections
+2.  reconnections
+3.  all clients connect with the same `clientid`.
```

Comment:
more tests:
4. while having a number of session registered, kill replicants see if core stands.
5. reconnect with/without clean-session
6. disconnects with session-disabled
7. disconnects with session-enabled